rfd/0163-vnet.md

```diff
@@ -0,0 +1,416 @@
+configuration or even having to know that Teleport is involved.
+
+## Why
+
+Teleport App Access has a pretty good UX for HTTP apps accessed through a web
+browser, but the experience isn't up to par outside of the browser.
+
+Today, Teleport users have two choices for accessing Teleport-protected apps and
+APIs with non-browser clients:
+
+1. Use a custom TLS certificate and private key with the client and connect to a
+   subdomain of the Teleport proxy, e.g. `curl --cert /Users/nic/.tsh/keys/one.teletest.private/nic-app/one.teletest.private/dumper-x509.pem --key /Users/nic/.tsh/keys/one.teletest.private/nic`.
+2. Proxy the app to a TCP listener on localhost, e.g. `tsh proxy app dumper --port 8888 & curl 127.0.0.1:8888`
+
+Both of these methods are fairly cumbersome, inflexible, and they may not even
+be possible with some clients, especially when needing to connect to many apps
+or using custom scripts or software that expects to be able to find these apps
+at a regular-old DNS address.
+Teleport VNet removes the need for manual management of client certs or tunnels.
+
+## Details
+
+Teleport VNet will run in the background on end-user devices (laptops, workstations).
+It will be integrated into Teleport Connect, and may also be available as a
+`tsh` command (TBD).
+
+It works by creating a TUN virtual network interface and configuring an IP route
+to send all outgoing IP packets in the VNet range through that interface.
+In the background, Teleport VNET will handle all incoming TCP connections to the
+TUN interface and proxy them to Teleport apps over authenticated tunnels, in
+much the same way that the `tsh proxy app` command works today.
+
+The VNET will also host a local DNS service serving custom domain names for all
+Teleport-protected apps, and directing the DNS queries for each app to an IP
+address in the VNet's virtual IP range assigned to the app in question.
+
+### Domain name assignment
+
+The default VNet domain name for each app will be `<app>.<proxy-address>.internal`.
+For an app named `api` in a Teleport cluster at `teleport.example.com` this
+would look like `api.teleport.example.com.internal`.
+
+There will be two ways to override this.
+
+The `spec.public_addr` field of the app resource already overrides the public
+address of the app for web and TCP app access.
+If the `public_addr` is set for an app, that address will be used in VNET, with
+`.internal` appended.
+An app with `public_addr: api.example.com`, would be accessed at
+`api.example.com.internal` in VNet.
+
+To override the address just for VNet, a new `spec.vnet_addr` field will be
+added on app resources.
+When this is used, the `internal` suffix will NOT automatically be added.
+An app with `vnet_addr: api.example.com` would be accessed at `api.example.com`
+in VNet.
+If `spec.vnet_addr` is not set it will default to `spec.public_addr`, which in
+turn defaults to `<app-name>.<proxy-public-address>`.
```

Comment:
I'll clarify a couple things and then answer all these questions

`app.spec.vnet_addr` will always be set and default to `app.spec.public_addr + ".internal"`

We will NOT try to handle every domain in a `vnet_addr` unless there is a custom DNS zone for it

Even if an app has an unreachable `vnet_addr` it can always be reached at `<app>.<proxy-address>.internal`

1. I believe we can do this without a cache, see the sequence under "DNS queries and IP address assignment", when a query comes in we can look up the app with a predicate search

2. We won't be handling anything in a custom DNS zone unless it is configured at the cluster level in a `vnet` resource under `spec.custom_dns_zones`. If an app has a `vnet_addr` with an address not in any custom DNS zone, we won't install the VNet nameserver for that zone. We will never need to change the host DNS configuration based on app changes alone, only when the `vnet` resource with the `custom_dns_zones` changes, maybe we can watch that or it's just fine that you'll have to relogin if you change that which should be infrequent

3. I don't think it's that much trickier. Most people will never have to know about it. Apps will always be reachable at `app.<proxy-address>.internal`, but if you really want to you can set `spec.vnet_addr` to set an alternative name. I thought defaulting `vnet_addr` to `public_addr+".internal"` might be convenient but it's probably not, I think I'll remove that idea and keep it empty
    - `bar.company.com.internal` wouldn't work in the first place because `bar.company.com` it is not a subdomain of the proxy address `teleport.example.com`. You could reach it at `foo.teleport.example.com.internal`, you'd need a custom DNS zone to reach it anywhere else, unless you set a simple alias like `fooalias.teleport.example.com.internal`

5. If an app has `public_addr` set to `foo.company.com` the `vnet_addr` will default to `foo.company.com.internal`. Without a custom DNS zone VNet will never receive queries for `company.com.internal`. Everything else under `.company.com` will keep working fine without VNet. If you want to access the app at `foo.company.com` you need to set `vnet_addr` to `foo.company.com` and create the custom DNS zone

Reply:
I reworked this section to drop the `public_addr` stuff, we'll ignore that, hopefully its more clear now that `<app>.<proxy-public-addr>.internal` will always word and `vnet_addr` can set an alternative

Reply:
It all makes sense, now that I understand that a cache is not needed for `vnet_addr` to work and that `vnet_addr` works only if there's a corresponding custom DNS zone.

There's still a potential for misconfiguration. We should be able to list apps which have `vnet_addr` set that doesn't match any custom DNS zone. Maybe something for v2.